### PR TITLE
volumemgr handle content tree parts

### PR DESCRIPTION
**Pull request by deitch**
Replaces the OCI image all-in-one tar download and then load into containerd at snapshot creation time, with downloading in parts.

Notable changes:

* downloads the index (if relevant), manifest, config and layers via downloader, then verifies each via verifier
* when all are verified, loads them into containerd
* `SnapshotPrepare()` has been changed to use the content already loaded into containerd

Based on @zed-rishabh's latest work on volumemgr

What does it not do well, or yet:

* does not load VMs into containerd. Given the size of them, and that they need media types (which are subject to discussion), I wanted to keep that separate. It also minimizes the changes
* it does not check if some element of content already exists in containerd before downloading and verifying. That is relatively simple, but should be a separate PR.
* does not remove the old verified files from `/persist/downloads/...` after loading into containerd. We didn't do it before, so I didn't want to change it here. We definitely can, and it is simple, in a future PR

Thus far, I have tested individually with downloads and verifications, but not kicking off actually ECOs. I will do that in the integration lab with the kind help of @eriknordmark, who offered.

Open questions for reviewers:

* load VMs as well: I recommend we leave that for a future PR
* remove old layers after verifying and loading into containerd: we should do that in a future PR
* are there too many `log.Infof()` statements? I had to put a lot in for debugging; would be good to get input from reviewers
* progress indication is weaker than it was before. I had gotten lots of advancement in the "tar downloader", and now we bypass it. It updates the volume, but only after each blob is downloaded. I recommend we let it go in this PR, and then update in a future one (before the next release).

Finally, there is one "bug". For reasons unknown (yet!), when downloading an image, larger layers appear briefly, start to download, then are marked as missing and disappear. I can recreate it reliably, but do not yet know why it is happening. The whole download then completes except for that layer, eventually kicking in again ~10-15 mins later. This obviously cannot get merged in with that bug, but having extra eyes on it would be very helpful.

I will load up logs into a gist and share it in a comment.

### CODE REVIEW

**Comment by deitch** [comment]
Re: the bug. When downloading the alpine:3.10 referenced in the alpha controller, there are three blobs to download:

* `21c83c5242199776c232920ddb58cfa2a46b17e42ed831ca9001c8dbc532d22d` - the binary layer, ~2.8MB
* `a143f3ba578f79e2c7b3022c488e6e12a35836cd4a6eb9e363d7f3a07d848590` - the manifest, 528B
* `be4e4bea2c2e15b403bb321562e78ea84b501fb41497472e91ecb41504e8a27c` - the config, ~1.6KB

The smaller layers download immediately, verify, and are good. The bigger layer starts to download - if I watch the downloads dir, I see `/persist/downloads/pending/21c83c5242199776c232920ddb58cfa2a46b17e42ed831ca9001c8dbc532d22d` briefly, but then it disappears, and doesn't come back for quite a while. 

I suspect some kind of race condition. All three layers get started, the small ones immediately kick off an `updateStatus`, the volumemgr checks the verify status of all three, then the download status of all three, just as the large layer is starting to download, and then deletes it. 

I pulled off the entire `/persist/rsyslog/syslog.txt`, and have it. The attached gist [here](https://gist.github.com/deitch/4b9c2f28d81acad851a4312d76e85ce5) has the results of `grep 21c83c5242199776c232920ddb58cfa2a46b17e42ed831ca9001c8dbc532d22d syslog.txt`.

If you follow it through, you can see it starting the download [here](https://gist.github.com/deitch/4b9c2f28d81acad851a4312d76e85ce5#file-layer-json-L20) at 11:55:49, lots of checking the status (I suspect because there were other updates that caused the whole volumemgr to check the statuses of all of its blobs), and then it has a "stat; no such file or directory" error [here](https://gist.github.com/deitch/4b9c2f28d81acad851a4312d76e85ce5#file-layer-json-L81) at 11:55:56, ~7 seconds later.

Any input would be appreciated.

**Comment by deitch** [comment]
FYI, there are some extra logging statements and even an `exec("ls","-l",localFile)`; these are for debugging purposes and will not be in the final commit.

File: pkg/pillar/cmd/volumemgr/blob.go [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+		// Nothing to do
+	case types.DOWNLOADING:
+		// Nothing to do
+	case types.DOWNLOADED:
+		// signal verifier to start if it hasn't already; add RefCount
+		if verifyBlob(ctx, objType, sv, blob) {
+			changed = true
+		}
+	}
+	if blob.WaitingForCerts {
```
Comment (eriknordmark) [inline pkg/pillar/cmd/volumemgr/blob.go:89]: I think we can do the whole waitingforcerts inside the verifier.
This if block is a no-op in any case.

File: pkg/pillar/cmd/volumemgr/blob.go [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+		log.Infof("Waiting for certs for blob %s", blob.Sha256)
+		return changed
+	}
+	log.Infof("Waiting for download for blob %s", blob.Sha256)
+	return changed
```
Comment (eriknordmark) [inline pkg/pillar/cmd/volumemgr/blob.go:94]: If the state became downloaded, this log message doesn't match.

File: pkg/pillar/cmd/volumemgr/blob.go [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+}
+
+// verifyBlob ensure a particular blob is being verified
```
Comment (eriknordmark) [inline pkg/pillar/cmd/volumemgr/blob.go:97]: It doesn't do that; it merely kicks the verifier.

File: pkg/pillar/cmd/volumemgr/blob.go [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+				blob.HasVerifierRef = true
+			}
+			//Wait for VerifyImageStatus to appear
+			return nil
+		}
+	} else {
+		log.Infof("Found verified blob status based on sha %s", blob.Sha256)
+		// If we don't already have a RefCount add one
+		// No need to checkCerts since we have a VerifyImageStatus
+		verifyBlob(ctx, objType, sv, blob)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/volumemgr/blob.go:165]: Here you need to update state based on VerifyImageStatus; verifyBlob doesn't do that AFAICT

File: pkg/pillar/cmd/volumemgr/handledownloader.go [hunk 5]
```diff
@@ -39,33 +38,27 @@ func AddOrRefcountDownloaderConfig(ctx *volumemgrContext, status types.ContentTr
...
 	} else {
-		log.Debugf("AddOrRefcountDownloaderConfig: add for %s",
-			status.ContentSha256)
+		log.Debugf("AddOrRefcountDownloaderConfig: add for %s", blob.Sha256)
 	}
-	name := status.RelativeURL
 
 	// where should the final downloaded file be?
-	locFilename := path.Join(types.DownloadDirname, status.ObjType, "pending", status.ContentID.String(), path.Base(name))
+	locFilename := path.Join(types.DownloadDirname, "pending", blob.Sha256)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/volumemgr/handledownloader.go:45]: A hacky workaround would be to add blob.Sha256 twice to the pathname. Alternatively, fix the offending code in the verifier,

File: pkg/pillar/cmd/volumemgr/blob.go (hunk 2 above)
Comment (deitch) [reply pkg/pillar/cmd/volumemgr/blob.go:94]: Good point. Copy-paste error. Fixing.

File: pkg/pillar/cmd/volumemgr/blob.go (hunk 3 above)
Comment (deitch) [reply pkg/pillar/cmd/volumemgr/blob.go:97]: Updated.

File: pkg/pillar/cmd/volumemgr/handledownloader.go (hunk 5 above)
Comment (deitch) [reply pkg/pillar/cmd/volumemgr/handledownloader.go:45]: This was a brilliant catch; thank you for finding it. I updated the offending code in verifier.

File: pkg/pillar/cmd/volumemgr/blob.go (hunk 4 above)
Comment (deitch) [reply pkg/pillar/cmd/volumemgr/blob.go:166]: Correct. `verifyBlob` checks for `HasVerifierRef`:

* has ref? do nothing
* doesn't have ref? add `VerifyImageConfig{}` to start the verifier

> update state based on VerifyImageStatus

It does. This is just checking the `VerifyStatus` and retrieving it. It is updated [here](https://github.com/lf-edge/eve/pull/1110/files#diff-eb3ec09d4c0f3b985be7625eded82029R106) for `ContentTree` and [here](https://github.com/lf-edge/eve/pull/1110/files#diff-f8d21b734fa414774d21d4219a2676b1R58) for `OldVolumeStatus`.

File: pkg/pillar/cmd/volumemgr/blob.go (hunk 1 above)
Comment (deitch) [reply pkg/pillar/cmd/volumemgr/blob.go:89]: Good point, as we discussed over Slack. Most of it was moved over; I missed this one. Fixed.

**Comment by deitch** [comment]
@eriknordmark I responded to all of your comments. Please take a look.

**Comment by zed-rishabh** [comment]
@deitch Gentle reminder: OldVolumeConfig/Status is going away in a few days.

**Comment by deitch** [comment]
Yup. But as long as it is there, I need to support it.

@zed-rishabh my tests thus far all used `OldVolume`; I just created an appinstance from alpine:3.10 and it did it. Is there a way for me to exercise `ContentTree`?

**Comment by zed-rishabh** [comment]
In master, we are creating both old and new volumes. New volumes are using the content tree and getting stored in ```persist/vault/volumes```. And with this PR https://github.com/lf-edge/eve/pull/1102, we are completely moving to the volume refs in application instance config.

**Comment by deitch** [comment]
This branch was completely rebased on master, so I’ll assume it is exercising it then.

**Comment by zed-rishabh** [comment]
Yes, you can also verify it from content tree and volume status in ```var/run/volumemgr/appImg.obj```

**Comment by deitch** [comment]
This has added the following:

* rebased on @zed-rishabh 's changes
* simplified the verifier dirs somewhat: removed the `/persist/downloads/appImg.obj/{verifier|verified}/<sha>/<sha>` dir/file structure in favour of a simpler `/persist/downloads/appImg.obj/{verifier|verified}/<sha>` (just the file). I want to remove the objType too, but that is more work and we will defer
* publishes `BlobStatus` to pubsub, to support the work @adarsh-zededa is doing with #1094 

Once integration tests pass, and we get some good eyes on it, this will be good to go.

Follow-in work:

* load VMs into containerd
* remove all of the transient files from `/persist/downloads/verified` once they are in containerd; this requires doing initial load on startup and delete updates
* more dynamic progress indication. I know how to do this, but I want it as a separate smaller PR

File: pkg/pillar/cmd/verifier/verifier.go [hunk 6]
```diff
@@ -1086,9 +1039,9 @@ func populateInitialStatusFromVerified(ctx *verifierContext,
 }
 
 func getVerifierDir(objType string) string {
-	return verifierBasePath + "/" + objType + "/" + "verifier"
+	return path.Join(verifierBasePath, "verifier", objType)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/verifier/verifier.go:1042]: Why swap the order?

File: pkg/pillar/cmd/verifier/verifier.go [hunk 7]
```diff
@@ -1086,9 +1039,9 @@ func populateInitialStatusFromVerified(ctx *verifierContext,
...
 }
 
 func getVerifiedDir(objType string) string {
-	return verifierBasePath + "/" + objType + "/" + "verified"
+	return path.Join(verifierBasePath, "verified", objType)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/verifier/verifier.go:1046]: Why swap the order?

Comment (deitch) [reply pkg/pillar/cmd/verifier/verifier.go:1046]: No good reason. This is an artifact of when I tried to remove `objType` entirely, and had to defer, so I put it back. I will fix that ordering.

Comment (deitch) [reply pkg/pillar/cmd/verifier/verifier.go:1046]: Fixed

File: pkg/pillar/cmd/verifier/verifier.go (hunk 6 above)
Comment (deitch) [reply pkg/pillar/cmd/verifier/verifier.go]: Reverted

**Pull request merged**